File: 백준/Gold/7576. 토마토/토마토.py
import sys
from collections import deque
input = sys.stdin.readline

m, n = map(int, input().split())
queue = deque()
graph = []

for _ in range(n):
    graph.append(list(map(int, input().split())))

for i in range(n):
    for j in range(m):
        if graph[i][j] == 1:
            queue.append((i,j))

def bfs():
    dx = [-1,0,1,0]
    dy = [0,-1,0,1]
    while queue:
        x,y = queue.popleft()
        for i in range(4):
            nx = x+dx[i]
            ny = y+dy[i]
            if 0<=nx<n and 0<=ny<m:
                if graph[nx][ny] == -1:
                    continue
                if graph[nx][ny] == 0:
                    queue.append((nx,ny))
                    graph[nx][ny] = graph[x][y]+1

bfs()
answ = 0
for i in graph:
    for j in i:
        if j==0:
            print(-1)
            exit(0)
    answ = max(answ,max(i))

print(answ-1)

File: 백준/Gold/7576. 토마토/토마토.java
import java.io.*;
import java.util.*;

public class Main {
    private static class Node {
        private int x;
        private int y;
        
        public Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
        
        public int getX() {
            return x;
        }
        public int getY() {
            return y;
        }
    }
    
    static int n;
    static int m;
    static int[][] gr;
    static Queue<Node> q = new LinkedList<>();
    static final int[] dx = new int[]{1, -1, 0, 0};
    static final int[] dy = new int[]{0, 0, 1, -1};
    
    private static int bfs() {
        while(!q.isEmpty()) {
            Node cur = q.remove();
            int x = cur.getX();
            int y = cur.getY();
            
            for(int d=0; d<4; d++) {
                int nx = x + dx[d];
                int ny = y + dy[d];
                
                if(nx<0 || nx>=n || ny<0 || ny>=m) continue;
                if(gr[nx][ny] == -1 || gr[nx][ny] >= 1) continue;
                if(gr[nx][ny]==0) {
                    gr[nx][ny] = gr[x][y] + 1;
                    q.add(new Node(nx, ny));    
                }
            }
        }
        
        int max = Integer.MIN_VALUE;
        for(int[] vg : gr) {
            for(int v : vg) {
                if(v == 0) return -1;
                else {
                    max = Math.max(max, v-1);
                }
            }
        }
        
        return max;
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());
        
        gr = new int[n][m];
        for(int i=0; i<n; i++) {
            gr[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        }
        
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++) {
                if(gr[i][j] == 1) {
                    q.add(new Node(i, j));
                }
            }
        }
        
        System.out.println(bfs());
    }   
}